---
title: "Sarcoid Microbiome Figures and Data"
author: "Erik Clarke"
date: "4/20/2017"
output: html_document
params: 
  data_fp: "/home/ecl/data/000_Sarcoidosis/099_Manuscript/data_files"
  figure_fp: "/home/ecl/data/000_Sarcoidosis/099_Manuscript/Figures"
  device: ""
---

```{r setup, error=FALSE}
knitr::opts_chunk$set(error=FALSE)
library(SarcoidMicrobiome)
library(ggbeeswarm)
library(forcats)
library(stringr)
library(ggplot2)
library(dplyr)
opts$data_fp <- params$data_fp
opts$figure_fp <- params$figure_fp

theme_set(
  theme_classic(base_size = 12) +
    theme(
      plot.title = element_text(hjust=0),
      axis.ticks = element_line(size=0.75), 
      axis.line.x = element_line(colour = 'black', size=0.6, linetype='solid'),
      axis.line.y = element_line(colour = 'black', size=0.6, linetype='solid'),
      strip.background = element_blank(),
      strip.text = element_text(face = "bold", hjust=0)
    )
)
```

```{r load-data}
a.16s <- LoadData("A", "16S")
a.its <- LoadData("A", "ITS")
b.16s <- LoadData("B", "16S")
b.its <- LoadData("B", "ITS")
c.16s <- LoadData("C", "16S")
c.its <- LoadData("C", "ITS")
c.vir <- LoadData("C", "virome")
de.16s <- LoadData("DE", "16S")
de.its <- LoadData("DE", "ITS")
de.wgs <- LoadData("DE", "WGS")

# Reconcile and anonymize all the sample names (e.g. subjects) between 
# 16S/ITS/WGS sequencings of a set

# BAL:
.c.vir.subjectid <- as.character(c.vir$s$OriginatingSampleID)
.c.vir.subjectid <- factor(str_replace(.c.vir.subjectid, "HUP", ""))
.bal.unified <- fct_unify(list(
  droplevels(c.16s$s$DerivingOriginalSampleID),
  droplevels(c.its$s$DerivingSampleID),
  .c.vir.subjectid))
c.16s$s$SubjectID <- fct_anon_deterministic(.bal.unified[[1]], prefix="C")
c.16s$agg <- left_join(c.16s$agg, c.16s$s)
c.its$s$SubjectID <- fct_anon_deterministic(.bal.unified[[2]], prefix="C")
c.its$agg <- left_join(c.its$agg, c.its$s)
c.vir$s$SubjectID <- fct_anon_deterministic(.bal.unified[[3]], prefix="C")
c.vir$agg <- left_join(c.vir$agg, c.vir$s)

# LN/A:
.a.all <- fct_unify(list(
  factor(a.16s$s$OriginatingSampleID),
  factor(a.its$s$OriginatingSampleID)
))
a.16s$s$SubjectID <- fct_anon_deterministic(.a.all[[1]], prefix="A")
a.16s$agg <- left_join(a.16s$agg, a.16s$s)
a.its$s$SubjectID <- fct_anon_deterministic(.a.all[[2]], prefix="A")
a.its$agg <- left_join(a.its$agg, a.its$s)

# LN/B:
.lnb.unified <- fct_unify(list(
  factor(make.names(as.character(b.16s$s$DerivingOriginalSampleID), allow_=FALSE)),
  factor(b.its$s$SampleID)
))
b.16s$s$SubjectID <- fct_anon_deterministic(.lnb.unified[[1]], prefix="B")
b.16s$agg <- left_join(b.16s$agg, b.16s$s)
b.its$s$SubjectID <- fct_anon_deterministic(.lnb.unified[[2]], prefix="B")
b.its$agg <- left_join(b.its$agg, b.its$s)
```

## Fig 1: Set A Barcharts
```{r a-16s-barcharts}
invisible(with(a.16s, {  
  agg %>%
    filter(SampleType %in% c("lymph_node", "paraffin", "extr_blank")) %>%
    mutate(SampleType = fct_recode(SampleType, "water"="extr_blank")) %>%
    mutate(SampleType = relevel(SampleType, "paraffin")) %>%
    MakeBarchartData() %>%
    PlotBarchart() %>%
    SplitBarcharts() %>%
    SaveBarcharts(
      file.path(opts$figure_fp, sprintf("Fig1A_Set%s_%s_%%s.pdf", sampleset, datatype)))
}))
```

```{r a-its-barcharts}
invisible(with(a.its, {
  # browser()
  agg %>%
    filter(SampleType %in% c("lymph_node", "paraffin", "extr_blank")) %>%
    mutate(SampleType = fct_recode(SampleType, "water"="extr_blank")) %>%
    mutate(SampleType = relevel(SampleType, "paraffin")) %>%
    ungroup() %>%
    mutate(
      MinRankOrder = SarcoidMicrobiome:::min_rank(., "Order")) %>%
    mutate(
      MinRankOrder = forcats::fct_recode(
        MinRankOrder, 
        Other="Fungi", 
        Other="uncultured fungus (phylum)", 
        Other="Indeterminate",
        Other="uncultured Acremonium (order)")) %>%
    MakeBarchartData(taxa.level="MinRankOrder") %>%
    PlotBarchart() %>%
    SplitBarcharts() %>%
    SaveBarcharts(
      file.path(opts$figure_fp, sprintf("Fig1B_Set%s_%s_%%s.pdf", sampleset, datatype)))
}))
```


```{r a-cladosporiaceae}
invisible(within(a.its, {
  # browser()
  
  .s <- s %>% select(SampleID, StudyGroup, SampleType, SubjectID)
  
  clado <- agg %>% 
    filter(!SampleType %in% c("extr_blank", "pcr_h2o")) %>%
    select(SampleID, Family, count) %>%
    filter(Family == "Cladosporiaceae") %>%
    count(SampleID, wt=count) %>%
    droplevels() %>%
    left_join(s) %>%
    select(SampleID, n, SampleType, SubjectID) %>%
    complete(SampleType, SubjectID, fill=list(n=0)) %>%
    filter(SampleType %in% c("paraffin", "lymph_node")) %>%
    select(-SampleID) %>%
    left_join(.s) %>%
    filter(!is.na(SampleID))
  
  .s2 <- .s %>% distinct(SubjectID, StudyGroup) %>% droplevels()
  
  clado.diff <- group_by(clado, SubjectID) %>%
    select(SubjectID, n, SampleType) %>%
    spread(SampleType, n) %>%
    mutate(delta=lymph_node-paraffin) %>%
    ungroup() %>%
    distinct(SubjectID, delta) %>%
    left_join(.s2) %>%
    mutate(delta.log10=sign(delta)*log10(abs(delta))) %>%
    mutate(delta.log10=ifelse(is.nan(delta.log10), 0, delta.log10)) %>%
    mutate(sign = delta.log10 > 0)
  
  p <- ggplot(clado.diff, aes(x=StudyGroup, y=delta.log10)) +
    geom_hline(yintercept=0, linetype=2) +
    geom_boxplot(fill=NA, size=0.5, fatten=4) +
    # scale_fill_manual(values=pcoa.fills, guide=FALSE) +
    # scale_color_manual(values=pcoa.colors, guide=FALSE) +
    # scale_color_manual("",
    #   values=c("TRUE"="black", "FALSE"="black"),
    #   labels=c("TRUE"="Greater in sample", "FALSE"="No greater in sample than env."),
    #   breaks=c("TRUE", "FALSE")) +
    scale_fill_manual("",
      values=c("TRUE"="#2166ac", "FALSE"="#d1e5f0"),
      labels=c("TRUE"="Greater in sample", "FALSE"="No greater in sample than env."),
      breaks=c("TRUE", "FALSE")) +
    geom_quasirandom(
      aes(fill=sign), shape=21, bandwidth = 0.1, width=0.3, size=2 ) +
    scale_y_continuous(
      expand=c(0,0),
      breaks=c(-5, -4, -3, -2, -1, 0, 1, 2, 3, 4, 5), 
      limits=c(-5,5), 
      labels = function(x) scales::comma(10^(abs(x)) * sign(x))
      ) +
    scale_x_discrete(
      labels=c("healthy"="Healthy", "sarcoidosis"="Sarcoidosis")) +
    labs(
      y="Sample - Environment (reads)", 
      x=NULL, 
      title=NULL) +
    theme_classic(base_size = 8) +
    theme(
      axis.text=element_text(color="black"),
      axis.line=element_line(size=1, color="black", linetype=1),
      axis.line.x=element_line(size=1, color="black", linetype=1),
      axis.line.y=element_line(size=1, color="black", linetype=1),
      panel.grid.minor=element_blank(),
      legend.position="bottom",
      legend.direction="vertical")

  plot(p)
  
  ggsave(
    file.path(opts$figure_fp, "Fig1C_SetA_Cladosporiaceae.pdf"),
    plot=p,
    device=cairo_failsafe, family="Helvetica",
    height=3, width=2.1)
}))
```

## Fig 2: Set B Barcharts

```{r b-16s-barcharts}
invisible(within(b.16s, {
  agg %>%
    filter(SampleType %in% c("lymph_node", "water")) %>%
    mutate(StudyGroup = fct_recode(StudyGroup, "extr_ctrl"="ext_ctrl")) %>%
    # mutate(SubjectID = ifelse(SampleType == "water", "B00", as.character(SubjectID))) %>%
    MakeBarchartData() %>%
    PlotBarchart() %>%
    SplitBarcharts() %>%
    SaveBarcharts(
      file.path(opts$figure_fp, sprintf("Fig2A_Set%s_%s_%%s.pdf", sampleset, datatype)),
      height=opts$barchart.height/1.5)
}))
```

```{r b-its-barcharts}
invisible(within(b.its, {
  agg %>%
    group_by(SubjectID) %>%
    filter(sum(count) > 0) %>%
    filter(SampleType %in% c("lymph_node", "extr_blank")) %>%
    mutate(SampleType = fct_recode(SampleType, "water"="extr_blank")) %>%
    ungroup() %>%
    mutate(
      MinRankOrder = SarcoidMicrobiome:::min_rank(., "Order")) %>%
    mutate(
      MinRankOrder = forcats::fct_recode(
        MinRankOrder, 
        Other="Fungi", 
        Other="uncultured fungus (phylum)", 
        Other="Indeterminate")) %>%
    MakeBarchartData(taxa.level = "MinRankOrder") %>%
    PlotBarchart() %>%
    SplitBarcharts() %>%
    SaveBarcharts(
      file.path(opts$figure_fp, sprintf("Fig2B_Set%s_%s_%%s.pdf", sampleset, datatype)),
      height=opts$barchart.height/1.5)
}))
```

## Figure 3: Set C Barcharts

```{r c-16s-barcharts}
invisible(within(c.16s, {
   agg %>%
    filter(SampleType %in% c("BAL", "prewash")) %>%
    mutate(SampleType = relevel(SampleType, "prewash")) %>%
    droplevels() %>%
    MakeBarchartData() %>%
    group_by(SubjectID, StudyGroup) %>% 
    complete(SampleType, fill = list(count=0)) %>%
    PlotBarchart(fill.values = bacterial.colors, x.labs = opts$bal.labels) %>%
    SplitBarcharts(groups = c("sarcoidosis", "healthy")) %>%
    SaveBarcharts(
      file.path(opts$figure_fp, sprintf("Fig3A_Set%s_%s_%%s.pdf", sampleset, datatype)))
}))
```

```{r c-its-barcharts}
invisible(within(c.its, {
   agg %>%
    filter(SampleType %in% c("BAL", "prewash")) %>%
    mutate(SampleType = relevel(SampleType, "prewash")) %>%
    droplevels() %>%
    MakeBarchartData() %>%
    group_by(SubjectID, StudyGroup) %>% 
    complete(SampleType, fill = list(count=0)) %>%
    group_by(SubjectID) %>%
    filter(sum(count) > 0) %>%
    PlotBarchart(fill.values = fungal.colors, x.labs = opts$bal.labels) %>%
    SplitBarcharts(groups = c("sarcoidosis", "healthy")) %>%
    SaveBarcharts(
      file.path(opts$figure_fp, sprintf("Fig3B_Set%s_%s_%%s.pdf", sampleset, datatype)))
}))
```

```{r c-virome-barcharts}
invisible(within(c.vir, {
  tech.contam <- c(
    "Enterobacteria phage M13", "Enterobacteria phage T7", 
    "Enterobacteria phage phiX174 sensu lato","Bacillus phage phi29", 
    "Pseudomonas phage phi6", "Shamonda virus", "Human herpesvirus 7", 
    "Human herpesvirus 6", "Cyprinid herpesvirus 3")
   agg %>%
     filter(LibraryType == "Genomiphi") %>%
     filter(SampleType %in% c("BAL", "prewash")) %>%
     filter(!Species %in% tech.contam) %>% 
     mutate(SampleType = relevel(SampleType, "prewash")) %>%
     droplevels() %>%
     MakeBarchartData(taxa.level = "Family") %>%
     group_by(SubjectID, StudyGroup) %>% 
     complete(SampleType, fill = list(count=0)) %>%
     group_by(SubjectID) %>%
     filter(sum(count) > 0) %>%
     PlotBarchart(fill.values = NULL, x.labs = opts$bal.labels) %>%
     SplitBarcharts(groups = c("sarcoidosis", "healthy")) %>%
     SaveBarcharts(
       file.path(opts$figure_fp, sprintf("Fig3C_Set%s_%s_%%s.pdf", sampleset, datatype)))
}))
```

## Figure 4: Sets D/E Barcharts

```{r de-16s-barcharts}
invisible(within(de.16s, {
  .s <- distinct(s, Sample, .keep_all = TRUE) %>% select(-SampleID)
  .agg <- agg %>% group_by(Sample, otu) %>%
    summarize(count=sum(count)) %>%
    group_by(Sample) %>%
    mutate(proportion=count/sum(count)) %>%
    left_join(.s) %>%
    left_join(md) %>%
    rename(SampleID=Sample) %>%
    mutate(SubjectID=SampleID) %>% droplevels()
  tmp <- .agg %>%
    MakeBarchartData() %>%
    PlotBarchart(
      x.col="SampleID", 
      x.labs=c("B"="BL", "Sal"="SL"),
      theme=opts$barchart.theme + theme(axis.text.y=element_text(family="mono"))) %>%
    SplitBarcharts(groups=c("Sarcoidosis", "Control")) %>%
    SaveBarcharts(
       file.path(opts$figure_fp, sprintf("Fig4A_Set%s_%s_%%s.pdf", sampleset, datatype)),
       height=opts$barchart.height/1.5)
}))
```

```{r de-its-barcharts}
invisible(within(de.its, {
  agg %>%
    mutate(SampleType = fct_recode(SampleType, "water"="extr_blank")) %>%
    mutate(SubjectID=SampleID) %>%
    mutate(
      MinRankOrder = SarcoidMicrobiome:::min_rank(., "Order")) %>%
    mutate(
      MinRankOrder = forcats::fct_recode(
        MinRankOrder, 
        Other="Fungi", 
        Other="uncultured fungus (phylum)", 
        Other="Indeterminate",
        Other="Coniochaetales")) %>%
    MakeBarchartData(taxa.level="MinRankOrder") %>%
    PlotBarchart(
      x.col="SubjectID",
      x.labs=c("Blank"="BL", "Saline"="SL", "Kveim"="KV", "Spleen1"="S1", "Spleen2"="S2", "Spleen3"="S3"),
      theme=opts$barchart.theme + theme(axis.text.y=element_text(family="mono"))) %>%
    SplitBarcharts(groups=c("Sarcoidosis", "Control")) %>%
    SaveBarcharts(
      file.path(opts$figure_fp, sprintf("Fig4B_Set%s_%s_%%s.pdf", sampleset, datatype)),
      height=opts$barchart.height/1.5)
  # browser()
}))
```

```{r de-wgs-barcharts}
invisible(within(de.wgs, {
  agg %>% 
    filter(Order != "Primates") %>%
    mutate(SubjectID=SampleID) %>%
    MakeBarchartData() %>%
    PlotBarchart(
      x.col="SampleID", 
      x.labs=c("B"="BL", "Sal"="SL"),
      theme=opts$barchart.theme + theme(axis.text.y=element_text(family="mono"))) %>%
    SplitBarcharts(groups=c("Sarcoidosis", "Control")) %>%
    SaveBarcharts(
       file.path(opts$figure_fp, sprintf("Fig4C_Set%s_%s_%%s.pdf", sampleset, datatype)),
       height=opts$barchart.height/1.5)
}))
```

## Figure S1: Set A Community Differences

```{r}

```

