---
title: "Sarcoid Microbiome Figures and Data"
author: "Erik Clarke"
date: "4/20/2017"
output: html_document
params: 
  data_fp: "/home/ecl/data/000_Sarcoidosis/099_Manuscript/data_files"
  figure_fp: "/home/ecl/data/000_Sarcoidosis/099_Manuscript/Figures"
---

```{r setup, error=FALSE}
knitr::opts_chunk$set(error=FALSE)
library(dplyr)
library(forcats)
library(stringr)
library(ggplot2)
library(SarcoidMicrobiome)

opts$data_fp <- params$data_fp
opts$figure_fp <- params$figure_fp

theme_set(
  theme_classic(base_size = 12) +
    theme(
      plot.title = element_text(hjust=0),
      axis.ticks = element_line(size=0.75), 
      axis.line.x = element_line(colour = 'black', size=0.6, linetype='solid'),
      axis.line.y = element_line(colour = 'black', size=0.6, linetype='solid'),
      strip.background = element_blank(),
      strip.text = element_text(face = "bold", hjust=0)
    )
)
```

```{r load-data}
a.16s <- LoadData("A", "16S")
a.its <- LoadData("A", "ITS")
b.16s <- LoadData("B", "16S")
b.its <- LoadData("B", "ITS")
c.16s <- LoadData("C", "16S")
c.its <- LoadData("C", "ITS")
c.vir <- LoadData("C", "virome")
de.16s <- LoadData("DE", "16S")
de.its <- LoadData("DE", "ITS")
de.wgs <- LoadData("DE", "WGS")

# Reconcile and anonymize all the sample names (e.g. subjects) between 
# 16S/ITS/WGS sequencings of a set

# BAL:
.c.vir.subjectid <- as.character(c.vir$s$OriginatingSampleID)
.c.vir.subjectid <- factor(str_replace(.c.vir.subjectid, "HUP", ""))
.bal.unified <- fct_unify(list(
  droplevels(c.16s$s$DerivingOriginalSampleID),
  droplevels(c.its$s$DerivingSampleID),
  .c.vir.subjectid))
c.16s$s$SubjectID <- fct_anon_deterministic(.bal.unified[[1]], prefix="C")
c.16s$agg <- left_join(c.16s$agg, c.16s$s)
c.its$s$SubjectID <- fct_anon_deterministic(.bal.unified[[2]], prefix="C")
c.its$agg <- left_join(c.its$agg, c.its$s)
c.vir$s$SubjectID <- fct_anon_deterministic(.bal.unified[[3]], prefix="C")
c.vir$agg <- left_join(c.vir$agg, c.vir$s)

# LN/A:
.a.all <- fct_unify(list(
  factor(a.16s$s$OriginatingSampleID),
  factor(a.its$s$OriginatingSampleID)
))
a.16s$s$SubjectID <- fct_anon_deterministic(.a.all[[1]], prefix="A")
a.16s$agg <- left_join(a.16s$agg, a.16s$s)
a.its$s$SubjectID <- fct_anon_deterministic(.a.all[[2]], prefix="A")
a.its$agg <- left_join(a.its$agg, a.its$s)

# LN/B:
.lnb.unified <- fct_unify(list(
  factor(make.names(as.character(b.16s$s$DerivingOriginalSampleID), allow_=FALSE)),
  factor(b.its$s$SampleID)
))
b.16s$s$SubjectID <- fct_anon_deterministic(.lnb.unified[[1]], prefix="B")
b.16s$agg <- left_join(b.16s$agg, b.16s$s)
b.its$s$SubjectID <- fct_anon_deterministic(.lnb.unified[[2]], prefix="B")
b.its$agg <- left_join(b.its$agg, b.its$s)
```

## Fig 1: Set A Barcharts
```{r a-16s-barcharts}
invisible(with(a.16s, {  
  p <- agg %>%
    filter(SampleType %in% c("lymph_node", "paraffin", "extr_blank")) %>%
    mutate(SampleType = fct_recode(SampleType, "water"="extr_blank")) %>%
    mutate(SampleType = relevel(SampleType, "paraffin")) %>%
    MakeBarchartData() %>%
    PlotBarchart(x.labs=ln.labels, fill.values=bacterial.colors) %>%
    SplitBarcharts() %>%
    SaveBarcharts(file.path(opts$figure_fp, "test_%s.pdf"))
}))
```

